### PR TITLE
add  `cache_index_and_filter_blocks_with_high_priority` option

### PR DESCRIPTION
**Pull request by UncP**
Signed-off-by: UncP <uncp.xu@gmail.com>

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Please add `cache_high_pri_pool_ratio` too, they are usually used together.

**Comment by UncP** [comment]
`cache_high_pri_pool_ratio`, rocksdb does not support this arg directly, only db_bench supports.

**Comment by zhangjinpeng1987** [comment]
Please explore the mechanism of `cache_high_pri_pool_ratio`.

**Comment by siddontang** [comment]
refer https://github.com/facebook/rocksdb/blob/8fc3de3c622896d49480a53eac0ebb20301a645b/tools/db_bench_tool.cc#L2107

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @huachaohuang

File: src/rocksdb_options.rs [hunk 1]
```diff
@@ -77,8 +77,8 @@ impl BlockBasedOptions {
         }
     }
 
-    pub fn set_lru_cache(&mut self, size: size_t) {
-        let cache = crocksdb_ffi::new_cache(size);
+    pub fn set_lru_cache(&mut self, size: size_t, num_shard_bits: c_int, strict_capacity_limit: c_uchar, high_pri_pool_ratio: c_double) {
```
Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: Please add comment to demonstrate how to set `num_shard_bits` and `high_pri_pool_ratio`.

Comment (UncP) [reply src/rocksdb_options.rs]: fixed

File: src/rocksdb_options.rs [hunk 2]
```diff
@@ -77,6 +77,11 @@ impl BlockBasedOptions {
         }
     }
 
+    // the recommanded num_shard_bits should be calculated like this:
```
Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: Also should explain the behavior when set num_shard_bits = -1.

File: src/rocksdb_options.rs [hunk 3]
```diff
@@ -77,8 +77,13 @@ impl BlockBasedOptions {
...
     }
 
-    pub fn set_lru_cache(&mut self, size: size_t) {
-        let cache = crocksdb_ffi::new_cache(size);
+    // the recommanded num_shard_bits should be calculated like this:
+    //     num_shard_bits = (size + 4194303) / 4194304 (4MB)
+    // for example, if your size is 4GB, num_shard_bits = (4 * 1024^3 + 4194303) / 4194304 = 10
+    // the recommanded strict_capacity_limit is false if your memory is sufficient
+    // the recommanded high_pri_pool_ratio should be 0.05 or 0.1
+    pub fn set_lru_cache(&mut self, size: size_t, num_shard_bits: c_int, strict_capacity_limit: c_uchar, high_pri_pool_ratio: c_double) {
```
Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: s/c_uchar/bool
s/c_double/f64

File: src/rocksdb_options.rs [hunk 4]
```diff
@@ -77,8 +77,13 @@ impl BlockBasedOptions {
         }
     }
 
-    pub fn set_lru_cache(&mut self, size: size_t) {
-        let cache = crocksdb_ffi::new_cache(size);
+    // the recommanded num_shard_bits should be calculated like this:
+    //     num_shard_bits = (size + 4194303) / 4194304 (4MB)
+    // for example, if your size is 4GB, num_shard_bits = (4 * 1024^3 + 4194303) / 4194304 = 10
```
Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: Also should explain the behavior when set num_shard_bits = -1.

File: src/rocksdb_options.rs (hunk 3 above)
Comment (UncP) [reply src/rocksdb_options.rs]: it won't pass compiling because c does not support bool type, it will cause an error when rust calls c functions.

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987 @huachaohuang

**Comment by huachaohuang** [comment]
LGTM, but I think the recommended `num_shard_bits` is controversial.

**Comment by UncP** [comment]
yes, i forgot to change it.

**Comment by zhangjinpeng1987** [comment]
Please fix ci.

**Comment by siddontang** [comment]
CI still failed @UncP

**Comment by siddontang** [comment]
CI still failed

**Pull request merged**